python/rikai/spark/sql/codegen/mlflow_registry.py

```diff
@@ -0,0 +1,307 @@
+            return getattr(mlflow, self.flavor).load_model(self.uri)
+        finally:
+            mlflow.set_tracking_uri(old_uri)
+
+    def _load_spec_dict(
+        self, uri: str, tags: dict, params: dict, extras=None
+    ) -> dict:
+        """Convert the Run into a ModelSpec
+
+        Parameters
+        ----------
+        uri: str
+            Rikai model reference
+        tags: dict
+            Tags from the Mlflow Run
+        params: dict
+            Params from the Mlflow Run
+        extras: dict, default None
+            Extra options passed in at model registration time
+
+        Returns
+        -------
+        spec: Dict[str, Any]
+        """
+        extras = extras or {}
+        spec = {
+            "version": _get_model_prop(
+                tags,
+                extras,
+                CONF_MLFLOW_SPEC_VERSION,
+                MlflowLogger._CURRENT_MODEL_SPEC_VERSION,
+            ),
+            "schema": _get_model_prop(tags, extras, CONF_MLFLOW_OUTPUT_SCHEMA),
+            "model": {
+                "flavor": _get_model_prop(
+                    tags, extras, CONF_MLFLOW_MODEL_FLAVOR
+                ),
+                "uri": uri,
+            },
+        }
+
+        # transforms
+        spec["transforms"] = {
+            "pre": _get_model_prop(
+                tags, extras, CONF_MLFLOW_PRE_PROCESSING, False
+            ),
+            "post": _get_model_prop(
+                tags, extras, CONF_MLFLOW_POST_PROCESSING, False
+            ),
+        }
+
+        # options
+        options = dict(params or {})  # for training
+        for key, value in tags.items():
+            key = key.lower().strip()
+            if key.startswith("rikai.option."):
+                sub_len = len("rikai.option.")
+                options[key[sub_len:]] = value
+        if len(options) > 0:
+            spec["options"] = options
+
+        return spec
+
+
+def _get_model_prop(
+    run_tags: dict,
+    extra_options: dict,
+    conf_name: str,
+    default_value: Any = None,
+    raise_if_absent: bool = True,
+) -> Any:
+    value = run_tags.get(
+        conf_name, extra_options.get(conf_name, default_value)
+    )
+    if not value and raise_if_absent:
+        raise ValueError(
+            (
+                "Please use rikai.mlflow.<flavor>.log_model after "
+                "training, or specify {} in CREATE MODEL OPTIONS"
+            ).format(conf_name)
+        )
+    return value
+
+
+def codegen_from_spec(
+    spark: SparkSession, spec: dict, name: Optional[str] = None
+) -> str:
+    """Generate code from an MLFlow runid
+
+    Parameters
+    ----------
+    spark : SparkSession
+        A live spark session
+    spec : dict
+        the model spec info dict
+    name : str
+        The name of the model in the catalog
+
+    Returns
+    -------
+    str
+        Spark UDF function name for the generated data.
+    """
+    udf = udf_from_spec(spec)
```

Comment:
is this method exact the same as in `FSRegistry`?

Reply:
`udf_from_spec` is a shared function (comes from base.py)
FileSystemRegistry reads from the spec file to generate the spec. MlflowRegistry will get attributes from tracking server to generate the spec dict.